### PR TITLE
Formal review of refactored sortData

### PR DESCRIPTION
**Pull request by ank266**
Refactored sortData and created unit tests for each new method created

### CODE REVIEW

File: src/__tests__/ParserComponent.test.tsx [hunk 1]
```diff
@@ -306,7 +309,7 @@ describe('R1 Tests\n', () => {
     await wrapper.instance().parse(fileEvent);
 
     // make sure alert was created
-    expect(windowSpy).toHaveBeenCalled();
+    // expect(windowSpy).toHaveBeenCalled();
```
Comment (braunsonm) [inline src/__tests__/ParserComponent.test.tsx]: Why? Or remove.

File: src/components/ParserComponent.tsx [hunk 2]
```diff
@@ -143,81 +144,95 @@ export default class ParserComponent extends React.Component<ParserInterface,
...
+   * Purpose: sorts the array of data
+   * @precondition data contains a column of valid numbers that
+     * can be interpreted as a date in the selected date format
+   * @postcondition the data stored in the array is sorted by some date column
+   * @param {Array} data: the array of data to sort
+     * @param {any} key: the column name by which dates must be sorted
+   * @return {boolean}: true if data is sorted correctly
+     * and false if it failed to sort
+   */
+    sortData(data: Array<object>, key:any): boolean {
```
Comment (braunsonm) [inline src/components/ParserComponent.tsx]: The key is a string isn't it? Why is it set as any here?

File: src/components/ParserComponent.tsx [hunk 3]
```diff
@@ -143,81 +144,95 @@ export default class ParserComponent extends React.Component<ParserInterface,
...
+      const formatString = this.state.formatString;
+      const initialData = this.state.data;
+      const keyInt = `${key}_num`;
+
+      TimSort.sort(data, function(a: any, b: any) {
+        let val: any;
+        if (!a.hasOwnProperty(keyInt)) {
```
Comment (braunsonm) [inline src/components/ParserComponent.tsx]: Why do you do this if check twice?

File: src/components/ParserComponent.tsx [hunk 4]
```diff
@@ -143,81 +144,95 @@ export default class ParserComponent extends React.Component<ParserInterface,
...
+        return (a[keyInt] - b[keyInt]);
+      });
+
+      this.setState(() => {
+        return {
+          data,
+        };
+      });
+      // state should be updated
+      assert.notStrictEqual(this.state.data, [],
```
Comment (braunsonm) [inline src/components/ParserComponent.tsx]: This assertion doesn't make much sense, it never was empty was it? Shouldn' t you be checking that it changed aka was sorted?

File: src/components/ParserComponent.tsx [hunk 5]
```diff
@@ -143,81 +144,95 @@ export default class ParserComponent extends React.Component<ParserInterface,
...
+    };
+
+    /**
+   * Purpose: sorts the array of data
+   * @precondition data contains a column of valid numbers that
+     * can be interpreted as a date in the selected date format
+   * @postcondition the data stored in the array is sorted by some date column
+   * @param {Array} data: the array of data to sort
+     * @param {any} key: the column name by which dates must be sorted
+   * @return {boolean}: true if data is sorted correctly
```
Comment (braunsonm) [inline src/components/ParserComponent.tsx]: Not really what it does though right? The return check only sees if the data is different. It doesn't verify the sort.

File: src/components/ParserComponent.tsx [hunk 6]
```diff
@@ -143,81 +144,95 @@ export default class ParserComponent extends React.Component<ParserInterface,
...
+      TimSort.sort(data, function(a: any, b: any) {
+        let val: any;
+        if (!a.hasOwnProperty(keyInt)) {
+          val = moment(a[key], formatString);
+          if (val.isValid()) {
+            a[keyInt] = val.valueOf();
+          } else {
+            a[keyInt] = -1;
+          }
+        }
```
Comment (mkoma808) [inline src/components/ParserComponent.tsx]: Osgood specifically referenced this as being something that should be abstracted into a function as it's just repeating code.

File: src/__tests__/ParserComponent.test.tsx [hunk 7]
```diff
@@ -436,6 +439,91 @@ describe('<ParserComponent /> Unit Tests', () => {
     });
   });
 
+  describe('lookForDateKey()', () => {
+    const wrapper = shallow(<ParserComponent prompt={'Select ' +
+    'a CSV file: '} fileType={FileType.csv}
+    onChange={function() {
+    }}/>);
+    const instance = wrapper.instance() as ParserComponent;
+    it('should return column name with the valid date in the first row' +
```
Comment (braunsonm) [inline src/__tests__/ParserComponent.test.tsx]: Can you add tests that have the dates in other formats? So we can be sure it respects the `formatString`?

File: src/__tests__/ParserComponent.test.tsx [hunk 8]
```diff
@@ -436,6 +439,91 @@ describe('<ParserComponent /> Unit Tests', () => {
...
+        {'Date': '2/31/1992'},
+        {'Date': '12/21/1992'}];
+      expect(instance.dataIsValid(testArray)).toBeTruthy();
+    });
+    it('should return false when given empty array', () => {
+      const testArray: {}[] = [];
+      try {
+        expect(instance.dataIsValid(testArray)).toThrow('File has no data');
+      } catch (e) {
+        console.log(e);
```
Comment (braunsonm) [inline src/__tests__/ParserComponent.test.tsx]: I don't understand this catch. You check if something was thrown in the previous call so this doesn't seem necessary.

File: src/__tests__/ParserComponent.test.tsx [hunk 9]
```diff
@@ -436,6 +439,91 @@ describe('<ParserComponent /> Unit Tests', () => {
...
+    });
+  });
+
+  describe('dataIsValid()', () => {
+    const wrapper = shallow(<ParserComponent prompt={'Select ' +
+    'a CSV file: '} fileType={FileType.csv}
+    onChange={function() {
+    }}/>);
+    const instance = wrapper.instance() as ParserComponent;
+    it('should return true when given array that has data ', () => {
```
Comment (braunsonm) [inline src/__tests__/ParserComponent.test.tsx]: What about a test when it's not True?

File: src/components/ParserComponent.tsx [hunk 10]
```diff
@@ -143,81 +144,95 @@ export default class ParserComponent extends React.Component<ParserInterface,
...
+     * that contains the valid date of given date format
+   */
+    lookForDateKey(data: Array<object>): any {
+      assert.notStrictEqual(data[0], null,
+          'sortData(): data[0] is null');
+      for (const [key, value] of Object.entries(data[0])) {
+        const date1 = moment(String(value), this.state.formatString);
+        if (moment(date1, this.state.formatString).isValid()) {
+          return key;
+        }
```
Comment (mkoma808) [inline src/components/ParserComponent.tsx]: If no date column is found, then this doesn't return anything or throw an error.

File: src/components/ParserComponent.tsx [hunk 11]
```diff
@@ -458,9 +473,12 @@ export default class ParserComponent extends React.Component<ParserInterface,
...
               assert.notStrictEqual(this.state.data, [],
                   'parseCsv(): this.state.data is empty' +
-              'but still calling inferTypes & sortData');
-              this.columnTypes = this.inferTypes(this.state.data);
-              this.sortData(this.state.data);
+                    'but still calling inferTypes & sortData');
+              if (this.dataIsValid(this.state.data)) {
+                this.columnTypes = this.inferTypes(this.state.data);
+                this.sortData(this.state.data,
+                    this.lookForDateKey(this.state.data));
```
Comment (mkoma808) [inline src/components/ParserComponent.tsx]: This result of `lookForDateKey` should be put into a variable.

File: src/components/ParserComponent.tsx (hunk 2 above)
Comment (mkoma808) [inline src/components/ParserComponent.tsx]: Should probably put an assert in here to make sure that `key` isn't null or underfined.

File: src/components/ParserComponent.tsx (hunk 6 above)
Comment (mkoma808) [reply src/components/ParserComponent.tsx]: To clarify, I mean that this:
`if (!a.hasOwnProperty(keyInt)) {
          val = moment(a[key], formatString);
          if (val.isValid()) {
            a[keyInt] = val.valueOf();
          } else {
            a[keyInt] = -1;
          }
        }`

File: src/components/ParserComponent.tsx [hunk 12]
```diff
@@ -190,76 +196,86 @@ export default class ParserComponent extends React.Component<ParserInterface,
...
-     *    if data does not contain year in some
-     *    dates but does in some it will sort lexicographically.
-     *    a csv has been uploaded, and the data is stored in an array.
-     * @postcondition the data stored in the array is sorted by some date column
-     * @param {Array} data: the array of data to sort
-     * @return {boolean}: array of sorted data
-     */
-    sortData(data: Array<object>): boolean {
+   * purpose: check if data is undefined and has more than 0 rows
+   * @precondition a csv has been uploaded and its data is stored in an array
```
Comment (braunsonm) [inline src/components/ParserComponent.tsx]: These docs are much less detailed and no longer have a post condition?

**Pull request merged**